keyboards/matrix/falcon_magnet/matrix.c

```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright 2023 astro
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include "apc.h"
+#include "analog.h"
+
+#define LEFT_KEY_PIN            A1
+#define RIGHT_KEY_PIN           A2
+#define LEFT_KEY_CHANNEL        ADC_CHANNEL_IN1
+#define RIGHT_KEY_CHANNEL       ADC_CHANNEL_IN2
+#define ADC_GRP1_NUM_CHANNELS   2
+#define ADC_GRP1_BUF_DEPTH      1
+
+static adcsample_t samples1[ADC_GRP1_NUM_CHANNELS * ADC_GRP1_BUF_DEPTH];
+
+/*
+ * ADC conversion group.
+ * Mode:        Linear buffer, 1 samples of 2 channel, SW triggered.
+ * Channels:    IN1&IN2.
+ */
+static const ADCConversionGroup adcgrpcfg1 = {
```

Comment:
Is there a reason that you aren't using the core adc/analog code here?

Reply:
as my understanding, the analog API in qmk was based on AVR, thus it lacks multi-channel ADC support. The multi-channele ADC can speed up the total matrix scan, it vital on those analog keyboards like EC and magnet.